### PR TITLE
[FLINK-30823][runtime] Enable speculative execution for typical built-in sinks

### PR DESCRIPTION
**Pull request by ifndef-SleePy**
## What is the purpose of the change

*This pull request enables speculative execution for some of typical built-in sinks*

## Brief change log

  - *Enable speculative execution for `DiscardingSink`*
  - *Enable speculative execution for `PrintSinkFunction`*
  - *Enable speculative execution for `PrintSink`*
  - *Enable speculative execution for `FileSink`*
  - *Enable speculative execution for `FileSystemOutputFormat`*
## Verifying this change

This change added tests:

  - *Added test case in `PartitionTempFileManagerTest`*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/table/PartitionTempFileManagerTest.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        fs.mkdirs(new Path(tmpPath.toUri() + "/task-1-attempt-0")); // invalid subtask index
+        fs.mkdirs(new Path(tmpPath.toUri() + "/.task-0-attempt-1")); // invisible dir
+        fs.mkdirs(new Path(tmpPath.toUri() + "/_SUCCESS")); // not a task dir
+
+        final List<Path> taskTmpPaths =
+                PartitionTempFileManager.listTaskTemporaryPaths(
+                        fs, new Path(tmpPath.toUri()), taskAttemptFilter);
+        final List<String> taskDirs =
+                taskTmpPaths.stream().map(Path::getName).collect(Collectors.toList());
+        assertThat(taskDirs).hasSize(1).contains("task-0-attempt-1");
```
Comment (zhuzhurk) [inline flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/table/PartitionTempFileManagerTest.java:56]: contains -> containsExactly

File: flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/FileSystemOutputFormat.java [hunk 2]
```diff
@@ -37,12 +38,17 @@
...
  *
  * @param <T> The type of the consumed records.
  */
 @Internal
-public class FileSystemOutputFormat<T> implements OutputFormat<T>, FinalizeOnMaster, Serializable {
+public class FileSystemOutputFormat<T>
+        implements OutputFormat<T>,
+                FinalizeOnMaster,
+                Serializable,
+                SupportsConcurrentExecutionAttempts {
```
Comment (zhuzhurk) [inline flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/FileSystemOutputFormat.java:51]: I know you have done e2e test. But it's better add some test cases as a guard for future changes to verify that the results are correct when speculative execution happens on FileSystemOutputFormat.

File: flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/sink/FileSink.java [hunk 3]
```diff
@@ -130,7 +131,8 @@
         implements StatefulSink<IN, FileWriterBucketState>,
                 TwoPhaseCommittingSink<IN, FileSinkCommittable>,
                 WithCompatibleState,
-                WithPreCommitTopology<IN, FileSinkCommittable> {
+                WithPreCommitTopology<IN, FileSinkCommittable>,
+                SupportsConcurrentExecutionAttempts {
```
Comment (zhuzhurk) [inline flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/sink/FileSink.java:135]: It's better add some test cases as a guard for future changes to verify that the results are correct when speculative execution happens on FileSink.

File: flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/PartitionTempFileManager.java [hunk 4]
```diff
@@ -121,13 +120,21 @@ private static String taskName(int task) {
...
     /** Returns task temporary paths in this checkpoint. */
-    public static List<Path> listTaskTemporaryPaths(FileSystem fs, Path basePath) throws Exception {
+    public static List<Path> listTaskTemporaryPaths(
+            FileSystem fs, Path basePath, BiPredicate<Integer, Integer> taskAttemptFilter)
+            throws Exception {
         List<Path> taskTmpPaths = new ArrayList<>();
 
         if (fs.exists(basePath)) {
             for (FileStatus taskStatus : fs.listStatus(basePath)) {
-                if (isTaskDir(taskStatus.getPath().getName())) {
```
Comment (zhuzhurk) [inline flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/PartitionTempFileManager.java:129]: The `isTaskDir(...)` method can be removed.

Comment (ifndef-SleePy) [reply flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/PartitionTempFileManager.java:129]: OK

File: flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/table/PartitionTempFileManagerTest.java (hunk 1 above)
Comment (ifndef-SleePy) [reply flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/table/PartitionTempFileManagerTest.java]: Nice catch.

**Comment by ifndef-SleePy** [comment]
@zhuzhurk How about opening another ticket for the it case? It might be a litter heavy for this pr. We can build a common it case framework to verify speculative execution for different connectors.

**Comment by zhuzhurk** [comment]
> @zhuzhurk How about opening another ticket for the it case? It might be a litter heavy for this pr. We can build a common it case framework to verify speculative execution for different connectors.

Sounds good to me.

**Review by zhuzhurk — APPROVED** [review]
LGTM.

**Pull request closed without merging**